README.md
# clj-cthun-client

CThun message client

https://github.com/puppetlabs/cthun-specifications


# Installation

The jar is distributed via the internal nexus server, to use it add
the following to your project.clj

    :dependencies [[puppetlabs/cthun-client "0.0.2"]]

    :repositories [["releases" "http://nexus.delivery.puppetlabs.net/content/repositories/releases/"]
                   ["snapshots" "http://nexus.delivery.puppetlabs.net/content/repositories/snapshots/"]]

# Usage example

```clojure
(ns example-client
  (:require [clojure.tools.logging :as log]
            [puppetlabs.cthun.client :as client]
            [puppetlabs.cthun.message :as message]))

(defn cnc-request-handler
  [conn request]
  (log/info "cnc handler got message" request)
  (let [response (-> (message/make-message)
                     (assoc :targets [(:sender request)]
                            :message_type "example/cnc_response")
                     (message/set-expiry 3 :seconds)
                     (message/set-json-data {:response "Hello world"
                                             :request (:id request)}))]
    (client/send! conn response))
  (log/info "cnc handler sent response"))

(defn default-request-handler
  [conn request]
  (log/info "Default handler got message" request))

;; connecting with handlers
(def conn (client/connect
           {:server "wss://localhost:8090/cthun/"
            :cert "test-resources/ssl/certs/0001_controller.pem"
            :private-key "test-resources/ssl/private_keys/0001_controller.pem"
            :cacert "test-resources/ssl/certs/ca.pem"
            :identity "cth://0001_controller/demo-client"
            :type "demo-client"}
           {"example/cnc_request" cnc-request-handler
            :default default-request-handler}))

;; sending messages
(client/send! conn
              (-> (message/make-message)
                  (message/set-expiry 3 :seconds)
                  (assoc :targets ["cth://*/demo-client"]
                         :message_type "example/any_schema")))

(client/send! conn
              (-> (message/make-message)
                  (message/set-expiry 3 :seconds)
                  (assoc :targets ["cth://*/demo-client"]
                         :message_type "example/cnc_request")
                  (message/set-json-data {:action "demo"})))

;; wait 5 seconds for things to resolve
(Thread/sleep (* 5 1000))

(client/close conn)
```


Changes.md
# 0.0.2

* Reworked for semantic changes in CTH-212

# 0.0.1

* Initial internal release


test/puppetlabs/cthun/client_test.clj
(ns puppetlabs.cthun.client-test
  (require [clojure.test :refer :all]
           [puppetlabs.cthun.client :refer :all]
           [puppetlabs.cthun.message :as message]))

(deftest session-association-message-test
  (let [message (session-association-message {:identity "cth://lolcathost/agent"})]
    (testing "it yields a message"
      (is (map? message)))
    (testing "message with the correct type"
      (is (= "http://puppetlabs.com/associate_request"
             (:message_type message))))
    (testing "message with the right sender"
      (is (= "cth://lolcathost/agent"
             (:sender message))))))

(deftest dispatch-message-test
  (with-redefs [puppetlabs.cthun.client/fallback-handler (fn [c m] "fallback")]

  (testing "should fall back to fallback-handler with no :default"
    (is (= "fallback"
           (dispatch-message {:handlers {}} {:message_type "foo"})))))
  (let [client {:handlers {"foo" (fn [c m] "foo")
                           :default (fn [c m] "default")}}]
    (testing "default handler should match when supplied"
      (is (= "foo"
             (dispatch-message client {:message_type "foo"})))
      (is (= "default"
             (dispatch-message client {:message_type "bar"}))))))


project.clj
(defn deploy-info
  [url]
  {:url url
   :username :env/nexus_jenkins_username
   :password :env/nexus_jenkins_password
   :sign-releases false})

(defproject puppetlabs/cthun-client "0.0.3-SNAPSHOT"
  :description "client library for cthun protocol"
  :url "https://github.com/puppetlabs/clj-cthun-client"
  :license {:name ""
            :url ""}

  :dependencies [[org.clojure/clojure "1.6.0"]
                 [org.clojure/tools.logging "0.3.1"]

                 [prismatic/schema "0.4.0"]

                 [puppetlabs/ssl-utils "0.8.0"]

                 [puppetlabs/cthun-message "0.1.0"]

                 ;; TODO(richardc) transitive from puppetlabs/cthun-message.  should not be needed
                 [clj-time "0.9.0"]

                 [stylefruits/gniazdo "0.4.0"]

                 ;; try+/throw+
                 [slingshot "0.12.2"]]

  :plugins [[lein-release "1.0.5"]]

  :repositories [["releases" "http://nexus.delivery.puppetlabs.net/content/repositories/releases/"]
                 ["snapshots"  "http://nexus.delivery.puppetlabs.net/content/repositories/snapshots/"]]

  :deploy-repositories [["releases" ~(deploy-info "http://nexus.delivery.puppetlabs.net/content/repositories/releases/")]
                        ["snapshots" ~(deploy-info "http://nexus.delivery.puppetlabs.net/content/repositories/snapshots/")]]
  :lein-release {:scm :git, :deploy-via :lein-deploy})


src/puppetlabs/cthun/client.clj
(ns puppetlabs.cthun.client
  (:require [clojure.tools.logging :as log]
            [gniazdo.core :as ws]
            [puppetlabs.cthun.message :as message]
            [puppetlabs.ssl-utils.core :as ssl-utils]
            [schema.core :as s])
  (:import  (java.nio ByteBuffer)
            (org.eclipse.jetty.websocket.client WebSocketClient)
            (org.eclipse.jetty.util.ssl SslContextFactory)))

;; schemas

(def ConnectParams
  "schema for connection parameters"
  {:server s/Str
   :cacert s/Str
   :cert s/Str
   :private-key s/Str
   :identity s/Str
   :type s/Str})

(def Handlers
  "schema for handler map.  String keys are data_schema handlers,
  keyword keys are special handlers (like :default)"
  {(s/either s/Str s/Keyword) (s/pred fn?)})

(def Client
  "schema for a client"
  (merge ConnectParams
         {:conn Object
          :websocket Object
          :handlers Handlers}))

;; private helpers for the ssl/websockets setup

(defn- make-ssl-context
  "Returns an SslContextFactory that does client authentication based on the client certificate named"
  ^SslContextFactory
  [params]
  (let [factory (SslContextFactory.)
        {:keys [cert private-key cacert]} params
        ssl-context (ssl-utils/pems->ssl-context cert private-key cacert)]
    (.setSslContext factory ssl-context)
    (.setNeedClientAuth factory true)
    factory))

(defn- make-websocket-client
  "Returns a WebSocketClient with the correct SSL context"
  ^WebSocketClient
  [params]
  (let [client (WebSocketClient. (make-ssl-context params))]
    (.start client)
    client))

(defn session-association-message
  [client]
  (-> (message/make-message)
      (message/set-expiry 3 :seconds)
      (assoc :sender (:identity client)
             :message_type "http://puppetlabs.com/associate_request"
             :targets ["cth:///server"])))

(defn fallback-handler
  "The handler to use when no handler matches"
  [client message]
  (log/debug "no handler for " message))

(defn dispatch-message
  [client message]
  (let [message-type (:message_type message)
        handlers (:handlers client)
        handler (or (get handlers message-type)
                    (get handlers :default)
                    fallback-handler)]
    (handler client message)))


(s/defn ^:always-validate send! :- s/Bool
  [client :- Client message :- message/Message]
  (ws/send-msg (:conn client)
               (message/encode (assoc message :sender (:identity client))))
  true)

;; TODO(richardc): the identity should be derived from the client
;; certificate and the connection type.

(s/defn ^:always-validate connect :- Client
  [params :- ConnectParams handlers :- Handlers]
  (let [client (promise)
        websocket (make-websocket-client params)
        conn (ws/connect (:server params)
                         :client websocket
                         :on-connect (fn [session]
                                       (log/debug "connected")
                                       (send! @client (session-association-message @client))
                                       (log/debug "sent login"))
                         :on-error (fn [error] (log/error "websockets error" error))
                         :on-close (fn [code message]
                                     (log/debug "websocket closed" code message))
                         :on-receive (fn [text]
                                       (log/debug "text message")
                                       (dispatch-message @client (message/decode (message/string->bytes text))))
                         :on-binary (fn [buffer offset count]
                                      (log/debug "bytes message" offset count)
                                      (log/debug "got " (message/decode buffer))
                                      (dispatch-message @client (message/decode buffer))))]
    (deliver client (assoc params :conn conn :websocket websocket :handlers handlers))
    @client))

(s/defn ^:always-validate close :- s/Bool
  "Close the connection"
  [client :- Client]
  (log/debug "Closing")
  (.stop (:websocket client))
  (ws/close (:conn client))
  true)


examples/README.clj
;; This is the example from the /README.md  Please keep me working.
;; to run, install the lein-exec plugin then:  lein exec -p examples/README.clj
(ns example-client
  (:require [clojure.tools.logging :as log]
            [puppetlabs.cthun.client :as client]
            [puppetlabs.cthun.message :as message]))

(defn cnc-request-handler
  [conn request]
  (log/info "cnc handler got message" request)
  (let [response (-> (message/make-message)
                     (assoc :targets [(:sender request)]
                            :message_type "example/cnc_response")
                     (message/set-expiry 3 :seconds)
                     (message/set-json-data {:response "Hello world"
                                             :request (:id request)}))]
    (client/send! conn response))
  (log/info "cnc handler sent response"))

(defn default-request-handler
  [conn request]
  (log/info "Default handler got message" request))

;; connecting with handlers
(def conn (client/connect
           {:server "wss://localhost:8090/cthun/"
            :cert "test-resources/ssl/certs/0006_controller.pem"
            :private-key "test-resources/ssl/private_keys/0006_controller.pem"
            :cacert "test-resources/ssl/certs/ca.pem"
            :identity "cth://0006_controller/demo-client"
            :type "demo-client"}
           {"example/cnc_request" cnc-request-handler
            :default default-request-handler}))

;; sending messages
(client/send! conn
              (-> (message/make-message)
                  (message/set-expiry 3 :seconds)
                  (assoc :targets ["cth://*/demo-client"]
                         :message_type "example/any_schema")))

(client/send! conn
              (-> (message/make-message)
                  (message/set-expiry 3 :seconds)
                  (assoc :targets ["cth://*/demo-client"]
                         :message_type "example/cnc_request")
                  (message/set-json-data {:action "demo"})))

;; wait 5 seconds for things to resolve
(Thread/sleep (* 5 1000))

(client/close conn)
